rfcs/003-workspaces.md

```diff
@@ -0,0 +1,70 @@
+# Workspaces
+
+# Summary
+
+Run persistent containers for users with web frontends, like VS Code and Jupyter Notebooks.
+
+# Motivation and background
+
+PrairieLearn currently allows code editing using an in-page ACE Editor and can compile and test code via external graders, which run take student code and instructor-provided test code and execute them in an instructor-defined container, returning the test results back to the student. While excellent for testing small code snippets, this is not a very flexible environment for writing and debugging more complex programs.
+
+This RFC proposes to give students persistent remote containers (called _Workspaces_) to work in, configured by instructors to provide a per-question environment with a specific set of compilers, debuggers, editors, etc. This remote container would be accessed via a web-based frontend, such as VS Code or Jupyter Notebooks.
+
+Goals:
+* Workspaces should be instructor-defined on a per-question basis.
+* Workspaces should launch from a PL question with a set of initial files in the home directory, which may be dynamically generated by the question `server.py` code.
+* The workspace should be accessible by the student via a web frontend (e.g., VS Code) that is served out of the workspace container.
+* The student should have complete freedom to use the workspace as a development environment, including compiling and executing code.
+* The files in the container home directory should be frequently autosaved to a persistent store, in case of container crashes (e.g., fork bombs).
+* At any time, the student should be able to trigger a "Grade" action, which will pull an instructor-defined set of files from the container and run the usual PL grading code (either internal or external graders).
+* Workspaces should auto-terminate after a period of inactivity (or force-terminated by the student), but should be able to be re-launched with the persistent home-directory files restored.
+
+# Proposed solution
+
+The server architecture has three conceptual components:
+1. Main PL web servers: render questions for the student with a "launch workspace" button in them.
+2. Manager servers: coordinate the launching of workspaces and proxy all traffic from the student browser through to the host machines.
+3. Host servers: run the actual containers.
+
+These three components are implemented within the main PL executable, but for deployment we can run different fleets of servers that use a `config` option to only turn on specific functionality.
```

Comment:
Wait, so the web/proxy/host servers are all the same "PL" server, just with respective settings enabled?

Reply:
Yep! This makes development easy, because you can run just one of them locally. For deployment we can choose whether to run them in separate autoscaling pools or combined in various combinations.